pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,508 @@
+			// Get the node name by destIP
+			entry := m.latencyStore.GetNodeIPLatencyEntryByKey(destIP)
+			if entry == nil {
+				klog.Warning("Failed to get node entry by destIP", "destIP", destIP)
+				continue
+			}
+
+			// Parse the ICMP message
+			var msg *icmp.Message
+			if isIPv4 {
+				msg, err = icmp.ParseMessage(ProtocolICMP, readBuffer[:n])
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+				if msg.Type != ipv4.ICMPTypeEchoReply {
+					klog.Warning("Failed to match ICMPTypeEchoReply")
+					continue
+				}
+			} else {
+				msg, err = icmp.ParseMessage(ProtocolIPv6ICMP, readBuffer)
+				if err != nil {
+					klog.ErrorS(err, "Failed to parse ICMP message")
+					continue
+				}
+				if msg.Type != ipv6.ICMPTypeEchoReply {
+					klog.Warning("Failed to match ICMPTypeEchoReply")
+					continue
+				}
+			}
+
+			echo, ok := msg.Body.(*icmp.Echo)
+			if !ok {
+				klog.Warning(nil, "Failed to assert type as *icmp.Echo")
+				continue
+			}
+
+			klog.InfoS("Recv ICMP message", "IP", destIP, "SeqID", entry.SeqID, "echo", echo)
+
+			// Parse the ICMP data
+			if entry.SeqID != uint32(echo.Seq) {
+				klog.Warning("Failed to match seqID", "entry.SeqID", entry.SeqID, "echo.Seq", echo.Seq)
+				continue
+			}
+
+			// Calculate the round-trip time
+			end := time.Now()
+			rtt := end.Sub(entry.LastSendTime)
+
+			// Update the latency store
+			m.latencyStore.UpdateNodeIPLatencyEntryByKey(destIP, &NodeIPLatencyEntry{
+				SeqID:           entry.SeqID,
+				LastSendTime:    entry.LastSendTime,
+				LastRecvTime:    end,
+				LastMeasuredRTT: rtt,
+			})
+		}
+	}
+}
+
+func (m *NodeLatencyMonitor) pingAll(ipv4Socket, ipv6Socket net.PacketConn) {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+	klog.InfoS("Start to ping all nodes")
+	for name, toIPs := range nodeIPs {
+		for _, toIP := range toIPs {
+			klog.InfoS("Start to ping node", "Node name", name, "Node IP", toIP)
+			if toIP.To4() != nil && ipv4Socket != nil {
+				if err := m.sendPing(ipv4Socket, toIP); err != nil {
+					klog.Warningf("Failed to send ICMP message to node, err: %#v, Node name: %#v, Node IP: %v", err, name, toIP)
+				}
+			} else if toIP.To16() != nil && ipv6Socket != nil {
+				if err := m.sendPing(ipv6Socket, toIP); err != nil {
+					klog.Warningf("Failed to send ICMP message to node, err: %#v, Node name: %#v, Node IP: %v", err, name, toIP)
+				}
+			} else {
+				klog.Warningf("Failed to send ICMP message to node, Node name: %#v, Node IP: %v", name, toIP)
+			}
+		}
+	}
+}
+
+func (m *NodeLatencyMonitor) report() {
+	// Print all connection status for debug
+	// It will be removed when collector is ready
+	klog.InfoS("Finish to ping all nodes")
+	entries := m.latencyStore.ListLatencies()
+
+	if len(entries) == 0 {
+		klog.Info("No latency data to report, skip reporting")
+		return
+	}
+
+	// Convert to NodeIPLatencyStats
+	nodeIPLatencyStats := &cpv1beta.NodeIPLatencyStat{
```

Comment:
One node has dual IPs?

One node only contains a key(name?) to store latencies in map in agg.
`map[nodename]NodeIPLatencyEntry`

Put a nodename to NodeLatencyEntry struct.